### ISSUE TITLE
Can't run the script, using libpython 2.7.9 instead of 3.x

### ISSUE THREAD

**Issue by stefanmaric** [tvdstaaij/telegram-json-backup#1]
Hi, I try to run the script but it exits with non-zero code.

```
Telegram-cli version 1.3.3, Copyright (C) 2013-2015 Vitaly Valtman
Telegram-cli comes with ABSOLUTELY NO WARRANTY; for details type `show_license'.
This is free software, and you are welcome to redistribute it
under certain conditions; type `show_license' for details.
Telegram-cli uses libtgl version 2.0.3
Telegram-cli includes software developed by the OpenSSL Project
for use in the OpenSSL Toolkit. (http://www.openssl.org/)
Telegram-cli uses libpython version 2.7.9
I: config dir=[/home/sam/.telegram-cli]
 *** Python Initialized
Fetching dialog list
User deleted user#124001284 deleted
User deleted user#124000739 deleted
User deleted user#124001089 deleted
User deleted user#124001261 deleted
User deleted user#71943474 deleted
User deleted user#124001773 deleted
User deleted user#124000145 deleted
User deleted user#58326045 deleted
User deleted user#5811636 deleted
User deleted user#36805467 deleted
User deleted user#7472832 deleted
User deleted user#5881334 deleted
User deleted user#14609232 deleted
User deleted user#40941857 deleted
User deleted user#46595462 deleted
User deleted user#18292829 deleted
User deleted user#12686656 deleted
User deleted user#8938203 deleted
> Selecting dialogs from list
SIGNAL received
h./bin/telegram-cli(print_backtrace+0x2f)[0x46f84f]
./bin/telegram-cli(termination_signal_handler+0x64)[0x46f8e4]
/lib/x86_64-linux-gnu/libc.so.6(+0x352f0)[0x7f16af5922f0]
/lib/x86_64-linux-gnu/libc.so.6(strlen+0x2a)[0x7f16af5e86fa]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyUnicodeUCS4_FromString+0x9)[0x7f16aff95609]
./bin/telegram-cli[0x48d129]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(_PyObject_GenericGetAttrWithDict+0xb5)[0x7f16b00b3845]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x146a)[0x7f16b00b9eaa]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x820)[0x7f16b0108f80]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x14daac)[0x7f16b0076aac]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43)[0x7f16b007f363]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_CallObjectWithKeywords+0x47)[0x7f16b01084d7]
./bin/telegram-cli(py_dialog_list_cb+0x1a9)[0x4893e9]
./bin/telegram-cli[0x497694]
./bin/telegram-cli(tglq_query_result+0x10f)[0x499a3f]
./bin/telegram-cli[0x49384e]
./bin/telegram-cli[0x4b4d9b]
/usr/lib/x86_64-linux-gnu/libevent-2.0.so.5(event_base_loop+0x7fc)[0x7f16b0f65a0c]
./bin/telegram-cli(net_loop+0xa4)[0x470e34]
./bin/telegram-cli(loop+0x195)[0x472205]
./bin/telegram-cli(main+0x2d3)[0x46e103]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f16af57da40]
./bin/telegram-cli(_start+0x29)[0x46e219]
```

On Ubuntu 15.04 and I do have Python 3.4.3 installed, but it seems tg is using 2.7. Any hint on how can I force it to use Python 3?

Thanks,

**Comment by tvdstaaij** [comment]
As far as I know this is not something you can change at runtime. You'll have to rebuild tg from scratch, linking with the correct Python version. Did you have Python 3 installed at the moment you ran `./configure` to build tg? This line suggests to me that newer Python versions are preferred by the configure script:
https://github.com/vysheng/tg/blob/0a1a3a9d7ac45f7bfa8a0dfc5fba0bc880207dbc/configure#L5950
I also have both Python versions on my Arch Linux system but configure picks Python 3. That may have something to do with Python 3 being the default on Arch, though.
If configure won't select Python 3 automatically you'll have to manually override the library version somehow. I have no idea how to do so, I suggest asking the tg folks.

**Comment by stefanmaric** [comment]
Hi @tvdstaaij, thank you for your response.

Digging into `tg`'s configure script, I noticed I actually had only the runtime installed, not libpython3.

I installed it and re-built `tg`. Now it is using python 3. But it still not working:

```
Telegram-cli version 1.3.3, Copyright (C) 2013-2015 Vitaly Valtman
Telegram-cli comes with ABSOLUTELY NO WARRANTY; for details type `show_license'.
This is free software, and you are welcome to redistribute it
under certain conditions; type `show_license' for details.
Telegram-cli uses libtgl version 2.0.3
Telegram-cli includes software developed by the OpenSSL Project
for use in the OpenSSL Toolkit. (http://www.openssl.org/)
Telegram-cli uses libpython version 3.4.3
I: config dir=[/home/sam/.telegram-cli]
 *** Python Initialized
Fetching dialog list
User deleted user#124001089 deleted
User deleted user#124001261 deleted
User deleted user#71943474 deleted
User deleted user#124000578 deleted
User deleted user#124000145 deleted
User deleted user#58326045 deleted
User deleted user#5811636 deleted
User deleted user#36805467 deleted
User deleted user#7472832 deleted
User deleted user#5881334 deleted
User deleted user#14609232 deleted
User deleted user#40941857 deleted
User deleted user#46595462 deleted
User deleted user#18292829 deleted
User deleted user#12686656 deleted
User deleted user#8938203 deleted
> Selecting dialogs from list
SIGNAL received
h./bin/telegram-cli(print_backtrace+0x2f)[0x46f89f]
./bin/telegram-cli(termination_signal_handler+0x64)[0x46f934]
/lib/x86_64-linux-gnu/libc.so.6(+0x352f0)[0x7f1b907142f0]
/lib/x86_64-linux-gnu/libc.so.6(strlen+0x2a)[0x7f1b9076a6fa]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(PyUnicode_FromString+0x9)[0x7f1b911521b9]
./bin/telegram-cli[0x48d169]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(_PyObject_GenericGetAttrWithDict+0x7c)[0x7f1b911300ac]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(PyEval_EvalFrameEx+0x39dc)[0x7f1b912510bc]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(PyEval_EvalCodeEx+0x883)[0x7f1b91255cf3]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(+0x12f3a9)[0x7f1b911da3a9]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(PyObject_Call+0x68)[0x7f1b9125a138]
/usr/lib/x86_64-linux-gnu/libpython3.4m.so.1.0(PyEval_CallObjectWithKeywords+0x47)[0x7f1b9124cab7]
./bin/telegram-cli(py_dialog_list_cb+0x1a9)[0x489439]
./bin/telegram-cli[0x4976a4]
./bin/telegram-cli(tglq_query_result+0x10f)[0x499a4f]
./bin/telegram-cli[0x49385e]
./bin/telegram-cli[0x4b4dab]
/usr/lib/x86_64-linux-gnu/libevent-2.0.so.5(event_base_loop+0x7fc)[0x7f1b921a3a0c]
./bin/telegram-cli(net_loop+0xa4)[0x470e84]
./bin/telegram-cli(loop+0x195)[0x472255]
./bin/telegram-cli(main+0x2d3)[0x46e153]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f1b906ffa40]
./bin/telegram-cli(_start+0x29)[0x46e269]
```

Any ideas? Could it be something related to vysheng/tgl version? I'm building from:

vysheng/tg@2052f4b381337d75e783facdbfad56b04dec1a9c (master)
vysheng/tgl@418f81be448fdebb04c3809795f8c160ee13280c (defined by submodule)

Do you have the details of your working setup?

Thanks!

**Comment by tvdstaaij** [comment]
You see, the `SIGNAL received` seems to be a generic catch-all panic handler for hard faults in tg's Python binding. And unfortunately there are some serious bugs in this binding that trigger such a fault condition, such as [tg#587](https://github.com/vysheng/tg/issues/587). I've already had to work around a few of them to get this script to work at all (just doing a `for...in` to iterate over the dialog list is enough to trigger a crash!).

I thought that your libpython version was a likely reason for the panic, but now that it still happens with Python 3 it is evident that there is something else causing it.

I will investigate and try to reproduce the issue. Best case, it's something that changed in a recent tg version and thus happens for everyone. Worst case, it's something specifically with your Telegram dialog list that triggers some unknown bug in the Python binding, as I've seen before. Even when I can reproduce it, and find out what causes it, I can't really _fix_ the bug, just try to avoid triggering it. This is pretty daunting and possibly even impossible if the bug is really bad, so I can't promise anything. But I'll try.

The Lua binding seems to be much more stable (in fact I think I had a Lua JSON backup script working in no time), but alas, it doesn't support fetching partial message backlogs as far as I know, meaning you can't go further back than a couple of thousand messages.

**Comment by tvdstaaij** [comment]
I did some debugging just now. If my hypothesis is right it turns out to be somewhere in between the worst and best case scenario. It's not something I can directly reproduce, but I think it's an old issue I'd already worked around -- except my workaround does not work for you. I tried a lot of things to see if I could make the workaround more generic, and I think I found a way. It's a pretty nasty issue, because just accessing or even checking the existence of the poisoned object property (`tgl.Peer.name`) triggers the crash.

Now that I've looked at this in more detail I have reason to believe that these poisoned objects belong to users you've chatted with before but have since deleted their Telegram account. This would also explain why my workaround works only for me: I have only one such user at the bottom of my contacts list, my guess is you have more than one.

The new workaround uses the absence of a peer's first name to predict the poisoning. Can you pull the new revision and check whether it works  for you?

**Comment by tvdstaaij** [comment]
While debugging this I discovered another related crash scenario -- messages in group chats sent by users who have deleted their account would also crash the script. This is now fixed with an additional commit.

Well actually, I'd noticed this one before but put off debugging it because of the complexity. But debugging your issue made me realize what was probably causing it. Thanks for reporting! I would never have found out otherwise.
